### PR TITLE
planner: fix wrong TableDual plans caused by comparing Binary and Bytes incorrectly (#23860)

### LINKED ISSUE #23846: The execution result of using TableDual plan is not correct
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
create table t(a varbinary(10),UNIQUE KEY(a));
insert into t values(0x00A4EEF4FA55D6706ED5);
select count(*) from t where a=0x00A4EEF4FA55D6706ED5; -- 1
desc select * from t where a=0x00A4EEF4FA55D6706ED5; -- TableDual_2 plan
select * from t where a=0x00A4EEF4FA55D6706ED5; -- Empty result
```
### 2. What did you expect to see? (Required)
0x00A4EEF4FA55D6706ED5
### 3. What did you see instead (Required)
Empty result
### 4. What is your TiDB version? (Required)

Release Version: v4.0.0-beta.2-2561-g9c75cfa4e
Edition: Community
Git Commit Hash: 9c75cfa4e2bbb854ab90562ecec0409f84a989b0
Git Branch: master
UTC Build Time: 2021-04-05 07:39:46
GoVersion: go1.15.1
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #23860 to release-5.0
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/23918
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/23918:release-5.0-af0982805940
```

---

### What problem does this PR solve?

Issue Number: close #23846 

Problem Summary: planner: fix wrong TableDual plans caused by comparing Binary and Bytes incorrectly

### What is changed and how it works?

planner: fix wrong TableDual plans caused by comparing Binary and Bytes incorrectly

### Check List 

Tests 

- Unit test

### Release note 

- planner: fix wrong TableDual plans caused by comparing Binary and Bytes incorrectly

### CODE REVIEW

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -3051,3 +3051,35 @@ func (s *testIntegrationSuite) TestIssue23736(c *C) {
 	// Should not use invisible index
 	c.Assert(tk.MustUseIndex("select /*+ stream_agg() */ count(1) from t0 where c > 10 and b < 2", "c"), IsFalse)
 }
+<<<<<<< HEAD
+=======
```
Comment (eurekaka) [inline planner/core/integration_test.go:3055]: @qw4990 please resolve conflicts.

**Comment by qw4990** [comment]
PTAL @time-and-fate @rebelice

**Review by time-and-fate — APPROVED** [review]
/lgtm

File: types/datum.go [hunk 2]
```diff
@@ -751,7 +751,7 @@ func (d *Datum) compareBinaryLiteral(sc *stmtctx.StatementContext, b BinaryLiter
 	case KindMaxValue:
 		return 1, nil
 	case KindString, KindBytes:
-		return CompareString(d.GetString(), b.ToString(), d.collation), nil
+		fallthrough // in this case, d is converted to Binary and then compared with b
 	case KindBinaryLiteral, KindMysqlBit:
 		return CompareString(d.GetBinaryLiteral().ToString(), b.ToString(), d.collation), nil
```
Comment (time-and-fate) [inline types/datum.go:756]: This line is different from the master branch, is this correct?

**Review by time-and-fate — APPROVED** [review]
/lgtm

**Comment by time-and-fate** [comment]
Seems this PR can only pass `TestIssue23846()` after https://github.com/pingcap/tidb/pull/24041 merged.

**Comment by zimulala** [comment]
@qw4990 
The test failed：
```
[2021-04-16T14:56:13.043Z] FAIL: integration_test.go:3137: testIntegrationSuite.TestIssue23846
[2021-04-16T14:56:13.043Z]
[2021-04-16T14:56:13.043Z] integration_test.go:3143:
[2021-04-16T14:56:13.043Z]     tk.MustQuery("select count(*) from t where a=0x00A4EEF4FA55D6706ED5").Check(testkit.Rows("1"))
[2021-04-16T14:56:13.043Z] /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:63:
[2021-04-16T14:56:13.043Z]     res.c.Assert(resBuff.String(), check.Equals, needBuff.String(), res.comment)
[2021-04-16T14:56:13.043Z] ... obtained string = "[0]\n"
[2021-04-16T14:56:13.043Z] ... expected string = "[1]\n"
[2021-04-16T14:56:13.043Z] ... sql:select count(*) from t where a=0x00A4EEF4FA55D6706ED5, args:[]
```

**Pull request merged**